README.md

```diff
@@ -79,6 +77,33 @@ The current implementation uses goroutines very liberally which means the query
 
 The current implementation creates a physical plan directly from the PromQL abstract syntax tree. Plan optimizations not yet implemented and would require having a logical plan as an intermediary step.
 
+## Distributed execution mode
+
+The engine supports a distributed mode where aggregations can be delegated to multiple remote engines, each responsible for an independent dataset. This mode is currently implemented through an optimizer which rewrites a query as a combination of multiple remote and one local aggregation. For example, when two remote engines are available, a query like:
+
+```
+sum(rate(http_request_total[4m]))
+```
+
+would be rewritten as
+
+```
+sum(
+  coalesce(
+    sum(rate(http_request_total[4m])) # remote engine 1
+    sum(rate(http_request_total[4m])) # remote engine 2
+  )
+)
+```
+
+The inner aggregations are forwarded to remote engines and the global result is completed in memory.
+
+An engine using the distributed mode can be created through the `NewDistributedEngine` function. The user is expected to pass an implementation of `RemoteEndpoints` with have a single `Engines()` method. When invoked, `Engines()` should return all remote engines that can be used for a single query. The `Engines()` method is called separately for each individual query which allows the `RemoteEndpoints` implementation to do continuous service discovery and inject engines as they become available.
+
+The interfaces used for remote execution can be found at [link](https://github.com/thanos-community/promql-engine/blob/main/api/remote.go). Note that the `RemoteEngine` interface has a `NewRangeQuery` method, similar to the one in the Prometheus [v1.QueryEngine](https://pkg.go.dev/github.com/prometheus/prometheus@v0.42.0/web/api/v1#QueryEngine) interface. It is up to the user of the library to implement this method as they see fit. An example implementation could be to forward the query to an HTTP `/api/v1/query_range` endpoint of a Prometheus instance. In Thanos, this method is implemented as a gRPC call to a Thanos Querier.
```

Comment:
Maybe using goDoc for all code references would be better?
```suggestion
The interfaces used for remote execution can be found in [api](https://pkg.go.dev/github.com/thanos-community/promql-engine/api) package. Note that the `RemoteEngine` interface has a `NewRangeQuery` method, similar to the one in the Prometheus [v1.QueryEngine](https://pkg.go.dev/github.com/prometheus/prometheus@v0.42.0/web/api/v1#QueryEngine) interface. It is up to the user of the library to implement this method as they see fit. An example implementation could be to forward the query to an HTTP `/api/v1/query_range` endpoint of a Prometheus instance. In Thanos, this method is implemented as a gRPC call to a Thanos Querier.
```

Reply:
Great point.